geopandas/_vectorized.py

```diff
@@ -0,0 +1,853 @@
+else:
+    type_mapping = None
+
+
+def _isna(value):
+    """
+    Check if scalar value is NA-like (None or np.nan).
+
+    Custom version that only works for scalars (returning True or False),
+    as `pd.isna` also works for array-like input returning a boolean array.
+    """
+    if value is None:
+        return True
+    elif isinstance(value, float) and np.isnan(value):
+        return True
+    else:
+        return False
+
+
+def _pygeos_to_shapely(geom):
+    if geom is None:
+        return None
+    elif pygeos.is_empty(geom) and pygeos.get_type_id(geom) == 0:
+        return shapely.wkt.loads("POINT EMPTY")
+    else:
+        return shapely.wkb.loads(pygeos.to_wkb(geom))
+
+
+def _shapely_to_pygeos(geom):
+    if geom is None:
+        return None
+    elif geom.is_empty and geom.geom_type == "Point":
+        return pygeos.from_wkt("POINT EMPTY")
+    else:
+        return pygeos.from_wkb(geom.wkb)
+
+
+def from_shapely(data):
+    """
+    Convert a list or array of shapely objects to an object-dtype numpy
+    array of validated geometry elements.
+
+    """
+    n = len(data)
+
+    out = []
+
+    for idx in range(n):
+        geom = data[idx]
+        if compat.USE_PYGEOS and isinstance(geom, pygeos.Geometry):
+            out.append(geom)
+        elif isinstance(geom, BaseGeometry):
+            if compat.USE_PYGEOS:
+                out.append(_shapely_to_pygeos(geom))
+            else:
+                out.append(geom)
+        elif hasattr(geom, "__geo_interface__"):
+            geom = shapely.geometry.asShape(geom)
+            if compat.USE_PYGEOS:
+                out.append(_shapely_to_pygeos(geom))
+            else:
+                out.append(geom)
+        elif _isna(geom):
+            out.append(None)
+        else:
+            raise TypeError("Input must be valid geometry objects: {0}".format(geom))
+
+    if compat.USE_PYGEOS:
+        return np.array(out, dtype=object)
+    else:
+        aout = np.empty(n, dtype=object)
+        aout[:] = out
+        return aout
+
+
+def to_shapely(data):
+    if compat.USE_PYGEOS:
+        out = np.empty(len(data), dtype=object)
+        out[:] = [_pygeos_to_shapely(geom) for geom in data]
+        return out
+    else:
+        return data
+
+
+def from_wkb(data):
+    """
+    Convert a list or array of WKB objects to a np.ndarray[geoms].
+    """
+    if compat.USE_PYGEOS:
+        return pygeos.from_wkb(data)
+
+    import shapely.wkb
+
+    n = len(data)
+
+    out = []
+
+    for idx in range(n):
+        geom = data[idx]
+        if geom is not None and len(geom):
+            geom = shapely.wkb.loads(geom)
+        else:
+            geom = None
+        out.append(geom)
+
+    aout = np.empty(n, dtype=object)
+    aout[:] = out
+    return aout
+
+
+def to_wkb(data):
+    if compat.USE_PYGEOS:
+        return pygeos.to_wkb(data)
+    else:
+        out = [geom.wkb if geom is not None else None for geom in data]
+        return np.array(out, dtype=object)
```

Comment:
Above and below is (repeatedly)
```
aout = np.empty(n, dtype=object)
aout[:] = out
```
while here directly `np.array(out, dtype=object)`. Is there a reason for the difference? Actually, is there a reason why everything else makes empty array first? It does not seem to be more performant.

Reply:
Reason is that if you have a single MultiGeometry, numpy will expand it to an array of its parts .. 
There were comments for this in the original code, but now that might be somewhere more down below. Will add it here as well

Reply:
I thought it will be related to MutliGeometry. Thanks!